libraries/AP_CINS/AP_CINS.h

```diff
@@ -0,0 +1,156 @@
+        quat.from_rotation_matrix(state.XHat.rot());
+        return quat.tofloat();
+    }
+    Location get_location() const {
+        Location loc = state.origin;
+        loc.offset(state.XHat.pos().x, state.XHat.pos().y);
+        loc.alt -= state.XHat.pos().z * 100;
+        return loc;
+    }
+    Vector3f get_velocity() const {
+        return state.XHat.vel().tofloat();
+    }
+    bool healthy(void) const {
+        return state.have_origin;
+    }
+    bool get_origin(Location &loc) {
+        loc = state.origin;
+        return state.have_origin;
+    }
+
+    static const struct AP_Param::GroupInfo var_info[];
+
+private:
+    void update_imu(const Vector3F &gyro_rads, const Vector3F &accel_mss, const ftype dt);
+    void update_gps(const Vector3F &pos, const Vector3F &vel, const ftype gps_dt);
+    void update_vector_measurement_cts(const Vector3F &measurement, const Vector3F& reference, const Vector2F &ref_base, const ftype& gain_R, const ftype& gain_V, const ftype& gain_gyr_bias, const ftype& gain_acc_bias, const ftype dt);
+    void update_attitude_from_compass();
+    bool init_yaw(void);
+    bool get_compass_yaw(ftype &yaw_rad, ftype &dt);
+    bool get_compass_vector(Vector3F &mag_vec, Vector3F &mag_ref, ftype &dt);
+
+    void compute_bias_update_imu(const SIM23& Gamma);
+    void saturate_bias(Vector3F& bias_correction, const Vector3F& current_bias, const ftype& saturation, const ftype& dt) const;
+
+    struct {
+        Vector3F accel;
+        Vector3F gyro;
+        Location origin;
+        bool have_origin;
+
+        //XHat and ZHat for CINS
+        Gal3F XHat; // State estimate containing attitude, velocity, position
+        SIM23 ZHat; // Auxiliary state for estimator maths
+
+        // Gyro Bias
+        Vector3F gyr_bias;
+        Vector3F gyr_bias_correction;
+        struct {
+            Matrix3F rot;
+            Matrix3F vel;
+            Matrix3F pos;
+        } gyr_bias_gain_mat;
+
+        // Accel Bias
+        Vector3F acc_bias;
+        Vector3F acc_bias_correction;
+        struct {
+            Matrix3F rot;
+            Matrix3F vel;
+            Matrix3F pos;
+        } acc_bias_gain_mat;
+    } state;
+
+    // Variables to be kept on the heap
+    struct {
+        Matrix3F A11;
+        Matrix3F A12;
+        Matrix3F A13;
+        Matrix3F A21;
+        Matrix3F A22;
+        Matrix3F A23;
+        Matrix3F A31;
+        Matrix3F A32;
+        Matrix3F A33;
+        Matrix3F M1Gyr;
+        Matrix3F M2Gyr;
+        Matrix3F M3Gyr;
+        Matrix3F M1Acc;
+        Matrix3F M2Acc;
+        Matrix3F M3Acc;
+        Matrix3F C11;
+        Matrix3F C12;
+        Matrix3F C13;
+        Matrix3F K11;
+        Matrix3F K21;
+        Matrix3F K31;
+        SIM23 Delta_SIM23;
+        SIM23 ZInv;
+    } heapVars;
+
+    // Delay variables required to handle delayed measurements.
+    struct {
+        ftype internal_time = 0.;
+        std::deque<std::pair<ftype,Gal3F>> stamped_inputs;
```

Comment:
This is problematic.

Should not be using standard library here

Reply:
fixed, replaced by ObjectBuffer